From cec6f4f51f22cd0324f255cb99a85eaa7f3d0d78 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Thu, 24 Dec 2015 09:51:31 +0100
Subject: [PATCH] video: mxc_edid: Adjust pixel timings

Our approach for fractional mode support requires the pixel timings
in the CEA mode table to describe the non-fractional modes. This
patch sets this for the modes that were defined otherwise.
Also it tweaks the rounding when calculating the real pixel clock
from the pixel time specified in the table.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/mxc/mxc_edid.c | 19 ++++++++++++-------
 include/video/mxc_edid.h     |  4 ++--
 2 files changed, 14 insertions(+), 9 deletions(-)

diff --git a/drivers/video/mxc/mxc_edid.c b/drivers/video/mxc/mxc_edid.c
index b3a6ee1..edfd6e8 100644
--- a/drivers/video/mxc/mxc_edid.c
+++ b/drivers/video/mxc/mxc_edid.c
@@ -38,20 +38,25 @@
 #define DPRINTK(fmt, args...)
 #endif
 
+/*
+ * Attention:	pixel times for the fractional mode entries must
+ *		specify the non-fractional frequency of that mode!
+ *		I.e. 60Hz instead of 59.94 and 24Hz instead of 23.98.
+ */
 const struct fb_videomode mxc_cea_mode[64] = {
 	/* #1: 640x480p@59.94/60Hz 4:3 */
 	[1] = {
-		NULL, 60, 640, 480, 39722, 48, 16, 33, 10, 96, 2, 0,
+		NULL, 60, 640, 480, 39682, 48, 16, 33, 10, 96, 2, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_ASPECT_4_3, 0,
 	},
 	/* #2: 720x480p@59.94/60Hz 4:3 */
 	[2] = {
-		NULL, 60, 720, 480, 37037, 60, 16, 30, 9, 62, 6, 0,
+		NULL, 60, 720, 480, 37000, 60, 16, 30, 9, 62, 6, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_ASPECT_4_3, 0,
 	},
 	/* #3: 720x480p@59.94/60Hz 16:9 */
 	[3] = {
-		NULL, 60, 720, 480, 37037, 60, 16, 30, 9, 62, 6, 0,
+		NULL, 60, 720, 480, 37000, 60, 16, 30, 9, 62, 6, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_ASPECT_16_9, 0,
 	},
 	/* #4: 1280x720p@59.94/60Hz 16:9 */
@@ -68,22 +73,22 @@ const struct fb_videomode mxc_cea_mode[64] = {
 	},
 	/* #6: 720(1440)x480iH@59.94/60Hz 4:3 */
 	[6] = {
-		NULL, 60, 1440, 480, 37037, 38, 114, 8, 31, 124, 6, 0,
+		NULL, 60, 1440, 480, 37000, 38, 114, 8, 31, 124, 6, 0,
 		FB_VMODE_INTERLACED | FB_VMODE_ASPECT_4_3, 0,
 	},
 	/* #7: 720(1440)x480iH@59.94/60Hz 16:9 */
 	[7] = {
-		NULL, 60, 1440, 480, 37037, 38, 114, 8, 31, 124, 6, 0,
+		NULL, 60, 1440, 480, 37000, 38, 114, 8, 31, 124, 6, 0,
 		FB_VMODE_INTERLACED | FB_VMODE_ASPECT_16_9, 0,
 	},
 	/* #8: 720(1440)x240pH@59.94/60Hz 4:3 */
 	[8] = {
-		NULL, 60, 1440, 240, 37108, 114, 38, 15, 4, 124, 3, 0,
+		NULL, 60, 1440, 240, 37000, 114, 38, 15, 4, 124, 3, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_ASPECT_4_3, 0,
 	},
 	/* #9: 720(1440)x240pH@59.94/60Hz 16:9 */
 	[9] = {
-		NULL, 60, 1440, 240, 37108, 114, 38, 15, 4, 124, 3, 0,
+		NULL, 60, 1440, 240, 37000, 114, 38, 15, 4, 124, 3, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_ASPECT_16_9, 0,
 	},
 	/* #14: 1440x480p@59.94/60Hz 4:3 */
diff --git a/include/video/mxc_edid.h b/include/video/mxc_edid.h
index 561fe61..2feed16 100644
--- a/include/video/mxc_edid.h
+++ b/include/video/mxc_edid.h
@@ -100,8 +100,8 @@ struct mxc_edid_cfg {
 };
 
 static inline unsigned long mxcPICOS2KHZ(u32 pixclock, u32 vmode) {
-	u32 x = (1000000000UL / (pixclock) * 1000 / ((vmode & FB_VMODE_FRACTIONAL) ? 1001 : 1000));
-	return x + ((1000000000UL % x) > (x / 2) ? 1 : 0);
+	return ((((vmode & FB_VMODE_FRACTIONAL) ?
+			(999000999UL*4) : (1000000000UL*4)) / pixclock) + 1) >> 2;
 }
 
 int mxc_edid_var_to_vic(struct fb_var_screeninfo *var);
-- 
1.9.1

